datastore/sqlite.nim

```diff
@@ -0,0 +1,238 @@
+import pkg/questionable
+import pkg/questionable/results
+import pkg/sqlite3_abi
+import pkg/upraises
+
+push: {.upraises: [].}
+
+# Adapted from:
+# https://github.com/status-im/nwaku/blob/master/waku/v2/node/storage/sqlite.nim
+
+type
+  AutoDisposed*[T: ptr|ref] = object
+    val*: T
+
+  DataProc* = proc(s: RawStmtPtr) {.closure.}
+
+  NoParams* = tuple # empty tuple
+
+  NoParamsStmt* = SQLiteStmt[NoParams, void]
+
+  RawStmtPtr* = ptr sqlite3_stmt
+
+  SQLite* = ptr sqlite3
+
+  SQLiteStmt*[Params, Res] = distinct RawStmtPtr
+
+proc sqlite3_bind_text(
+  pstmt: ptr sqlite3_stmt,
+  param: cint,
+  value: pointer,
+  n: cint,
+  dispose: proc (v: pointer) {.cdecl.}): cint {.importc, cdecl.}
+
+proc bindParam(
+  s: RawStmtPtr,
+  n: int,
+  val: auto): cint =
+
+  when val is openarray[byte]|seq[byte]:
+    if val.len > 0:
+      try:
+        # `SQLITE_TRANSIENT` "indicate[s] that the object is to be copied prior
+        # to the return from sqlite3_bind_*(). The object and pointer to it
+        # must remain valid until then. SQLite will then manage the lifetime of
+        # its private copy."
+        sqlite3_bind_blob(s, n.cint, unsafeAddr val[0], val.len.cint,
+          SQLITE_TRANSIENT)
+      except Exception as e:
+        raise (ref Defect)(msg: $e.msg)
+    else:
+      sqlite3_bind_blob(s, n.cint, nil, 0.cint, nil)
+  elif val is int32:
+    sqlite3_bind_int(s, n.cint, val)
+  elif val is uint32:
+    sqlite3_bind_int(s, int(n).cint, int(val).cint)
+  elif val is int64:
+    sqlite3_bind_int64(s, n.cint, val)
+  elif val is float64:
+    sqlite3_bind_double(s, n.cint, val)
+  elif val is string:
+    try:
+      # `-1` implies string length is num bytes up to first null-terminator;
+      # `SQLITE_TRANSIENT` "indicate[s] that the object is to be copied prior
+      # to the return from sqlite3_bind_*(). The object and pointer to it must
+      # remain valid until then. SQLite will then manage the lifetime of its
+      # private copy."
+      sqlite3_bind_text(s, n.cint, val.cstring, -1.cint, SQLITE_TRANSIENT)
+    except Exception as e:
```

Comment:
Current versions of nim allow you to handle system level exceptions (unfortunately), `Exception` is the base class of all fatal and non-fatal exceptions. 

Annotating the module/procedure with `raises: Defect` (enabled through `push: {.upraises: [].}` here) will signal the compiler to allow propagating only `Defect`. 

Most likely, something in the guts of sqlite raises an `Exception` which gets flagged by the `upraises/raises` mechanism and subsequently fails to compile. The only way to guarantee safety in such situations is to remap `Exception` into `Defect` and propagate, which should eventually "panic" the process thus avoiding it to continue execution in a potentially inconsistent state.

Reply:
Now, I'm not sure if this is actually the case @michaelsbradleyjr, if the `C` function is not raising anything, there is no reason to do the wrapping.

Reply:
Th need for `try/except` here was a bit of a mystery. I introduced it only because, when passing `SQLITE_TRANSIENT` instead of `nil`, the compiler failed with error message: `can raise an unlisted exception: Exception`.

It would be nice to have an explanation/understanding of what was happening, but with changes to [nim-sqlite3-abi](https://github.com/arnetheduck/nim-sqlite3-abi) that landed earlier today ([fda455cfea2df707dde052034411ce63de218453](https://github.com/arnetheduck/nim-sqlite3-abi/commit/fda455cfea2df707dde052034411ce63de218453)) it's no longer necessary to define `proc sqlite3_bind_text` in `datastore/sqlite.nim` and it's no longer necessary to wrap with `try/except` when passing `SQLITE_TRANSIENT`.

Reply:
Yep, that makes sense, probably missing `Defect` annotations or some such.

Reply:
I bet that without an annotation, `SQLITE_TRANSIENT = cast[callback](-1)` was considered as real callback that can raise any exception